include/experimental/__p1684_bits/mdarray.hpp

```diff
@@ -0,0 +1,484 @@
+    /* requires */ (container_is_array<container_type>::value)
+  ) : map_(m), ctr_()
+  { }
+
+  // Constructors from container
+  MDSPAN_TEMPLATE_REQUIRES(
+    class... SizeTypes,
+    /* requires */ (
+      _MDSPAN_FOLD_AND(_MDSPAN_TRAIT(is_convertible, SizeTypes, size_type) /* && ... */) &&
+      _MDSPAN_TRAIT(is_constructible, extents_type, SizeTypes...) &&
+      _MDSPAN_TRAIT(is_constructible, mapping_type, extents_type)
+    )
+  )
+  MDSPAN_INLINE_FUNCTION
+  explicit constexpr mdarray(const container_type& ctr, SizeTypes... dynamic_extents)
+    : map_(extents_type(dynamic_extents...)), ctr_(ctr)
+  { assert(ctr.size() >= map_.required_span_size()); }
+
+
+  MDSPAN_FUNCTION_REQUIRES(
+    (MDSPAN_INLINE_FUNCTION constexpr),
+    mdarray, (const container_type& ctr, const extents_type& exts), ,
+    /* requires */ (_MDSPAN_TRAIT(is_constructible, mapping_type, extents_type))
+  ) : map_(exts), ctr_(ctr)
+  { assert(ctr.size() >= map_.required_span_size()); }
+
+  constexpr mdarray(const container_type& ctr, const mapping_type& m)
+    : map_(m), ctr_(ctr)
+  { assert(ctr.size() >= map_.required_span_size()); }
+
+
+  // Constructors from container
+  MDSPAN_TEMPLATE_REQUIRES(
+    class... SizeTypes,
+    /* requires */ (
+      _MDSPAN_FOLD_AND(_MDSPAN_TRAIT(is_convertible, SizeTypes, size_type) /* && ... */) &&
+      _MDSPAN_TRAIT(is_constructible, extents_type, SizeTypes...) &&
+      _MDSPAN_TRAIT(is_constructible, mapping_type, extents_type)
+    )
+  )
+  MDSPAN_INLINE_FUNCTION
+  explicit constexpr mdarray(container_type&& ctr, SizeTypes... dynamic_extents)
+    : map_(extents_type(dynamic_extents...)), ctr_(std::move(ctr))
+  { assert(ctr_.size() >= map_.required_span_size()); }
+
+
+  MDSPAN_FUNCTION_REQUIRES(
+    (MDSPAN_INLINE_FUNCTION constexpr),
+    mdarray, (container_type&& ctr, const extents_type& exts), ,
+    /* requires */ (_MDSPAN_TRAIT(is_constructible, mapping_type, extents_type))
+  ) : map_(exts), ctr_(std::move(ctr))
+  { assert(ctr_.size() >= map_.required_span_size()); }
+
+  constexpr mdarray(container_type&& ctr, const mapping_type& m)
+    : map_(m), ctr_(std::move(ctr))
+  { assert(ctr_.size() >= map_.required_span_size()); }
+
+
+
+  MDSPAN_TEMPLATE_REQUIRES(
+    class OtherElementType, class OtherExtents, class OtherLayoutPolicy, class OtherContainer,
+    /* requires */ (
+      _MDSPAN_TRAIT(is_constructible, mapping_type, typename OtherLayoutPolicy::template mapping<OtherExtents>) &&
+      _MDSPAN_TRAIT(is_constructible, container_type, OtherContainer)
+    )
+  )
+  MDSPAN_INLINE_FUNCTION
+  constexpr mdarray(const mdarray<OtherElementType, OtherExtents, OtherLayoutPolicy, OtherContainer>& other)
+    : map_(other.mapping()), ctr_(other.container())
+  {
+    static_assert(is_constructible<extents_type, OtherExtents>::value, "");
+  }
+
+  // Constructors for container types constructible from a size and allocator
+  MDSPAN_TEMPLATE_REQUIRES(
+    class Alloc,
```

Comment:
As is, I suspect these constructors are under constrained. Constructing an `mdarray<int, ..., std::vector<int>>` with an `extents_type` and a `float` (as an example), would satisfy the requirements of this constructor. That causes the resulting vector to have the appropriate number of elements, each initialized the the value of the (truncated) `float`. Not the worst thing to possibly happen here, but I've not exhaustive checked all contructors.

Reply:
It occurred to me today `requires uses_allocator_v<container_type, Alloc>` or equivalent for pre-C++20 should be sufficient for constraining the allocator constructors.